code/parachain/frame/staking-rewards/src/lib.rs

```diff
@@ -852,198 +929,198 @@ pub mod pallet {
+				let share_asset_id =
+					RewardPools::<T>::try_mutate(stake.staked_asset_id, |rewards_pool| {
+						let rewards_pool =
+							rewards_pool.as_mut().ok_or(Error::<T>::RewardsPoolNotFound)?;
 
-					Self::collect_rewards(
-						rewards_pool,
-						&mut stake,
-						&owner,
-						is_early_unlock,
-						keep_alive,
-					)?;
+						Self::collect_rewards(
+							rewards_pool,
+							&mut stake,
+							&owner,
+							is_early_unlock,
+							keep_alive,
+						)?;
 
-					rewards_pool.claimed_shares =
-						rewards_pool.claimed_shares.safe_add(&stake.share)?;
+						rewards_pool.claimed_shares =
+							rewards_pool.claimed_shares.safe_add(&stake.share)?;
 
-					Ok::<_, DispatchError>((rewards_pool.asset_id, rewards_pool.share_asset_id))
-				})?;
+						Ok::<_, DispatchError>(rewards_pool.share_asset_id)
+					})?;
 
-			// REVIEW(benluelo): Make this logic a method on Stake
-			let stake_with_penalty = if is_early_unlock {
-				(Perbill::one() - stake.lock.unlock_penalty).mul_ceil(stake.stake)
-			} else {
-				stake.stake
-			};
+				// REVIEW(benluelo): Make this logic a method on Stake
+				let staked_amount_returned_to_staker = if is_early_unlock {
+					(Perbill::one() - stake.lock.unlock_penalty).mul_ceil(stake.amount)
+				} else {
+					stake.amount
+				};
 
-			let fnft_asset_account =
-				T::FinancialNft::asset_account(fnft_collection_id, fnft_instance_id);
+				let fnft_asset_account =
+					T::FinancialNft::asset_account(fnft_collection_id, fnft_instance_id);
 
-			T::Assets::remove_lock(T::LockId::get(), asset_id, &fnft_asset_account)?;
-			T::Assets::remove_lock(T::LockId::get(), share_asset_id, &fnft_asset_account)?;
-			T::Assets::transfer(
-				asset_id,
-				&fnft_asset_account,
-				&owner,
-				stake_with_penalty,
-				keep_alive,
-			)?;
-
-			Stakes::<T>::remove(fnft_collection_id, fnft_instance_id);
+				T::Assets::remove_lock(
+					T::LockId::get(),
+					stake.staked_asset_id,
+					&fnft_asset_account,
+				)?;
+				T::Assets::remove_lock(T::LockId::get(), share_asset_id, &fnft_asset_account)?;
+				T::Assets::transfer(
+					stake.staked_asset_id,
+					&fnft_asset_account,
+					&owner,
+					staked_amount_returned_to_staker,
+					keep_alive,
+				)?;
 
-			// Burn the financial NFT and the shares it holds
-			T::Assets::burn_from(asset_id, &fnft_asset_account, stake.stake - stake_with_penalty)?;
-			T::Assets::burn_from(share_asset_id, &fnft_asset_account, stake.stake)?;
-			T::FinancialNft::burn(fnft_collection_id, fnft_instance_id, Some(who))?;
+				// Burn the financial NFT and the shares it holds
+				if is_early_unlock {
+					// If there is no penalty then there is nothing to burn as it will all have been
+					// transferred back to the staker. burn_from isn't a noop if the amount to burn
+					// is 0, hence the check
+					T::Assets::burn_from(
+						stake.staked_asset_id,
+						&fnft_asset_account,
+						stake.amount.defensive_saturating_sub(staked_amount_returned_to_staker),
```

Comment:
SafeSub?

Reply:
`staked_amount_returned_to_staker` is always <= stake.amount, so this should never fail - I thought I had added a comment explaining that but I must have forgot. I will add an explanation :+1:

Reply:
So the slashed shares should be sent to the treasury?

Reply:
NOT "shares" (shares are not slashed) but the staked asset "amounts" that slashed are supposed to go to treasury